File: backend/app/api/v1/endpoints/estructura_eia.py
"""
Endpoints de Estructura EIA (Fase 2).

Permite generar y gestionar la estructura del EIA para proyectos
que requieren Estudio de Impacto Ambiental.
"""
import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.services.estructura_eia.service import (
    EstructuraEIAService,
    get_estructura_eia_service,
)
from app.schemas.estructura_eia import (
    EstructuraEIAResponse,
    EstructuraEIAResumen,
    CapituloConEstado,
    CapituloEIAResponse,
    PASConEstado,
    AnexoRequerido,
    ComponenteLineaBaseResponse,
    ComponenteLineaBaseEnPlan,
    EstimacionComplejidad,
    ActualizarCapituloRequest,
    ActualizarPASRequest,
    ActualizarAnexoRequest,
    ResumenEstructuraParaAsistente,
)

logger = logging.getLogger(__name__)

router = APIRouter()


# =============================================================================
# Dependencias
# =============================================================================

def get_servicio(db: AsyncSession = Depends(get_db)) -> EstructuraEIAService:
    """Obtiene instancia del servicio de estructura EIA."""
    return get_estructura_eia_service(db)


# =============================================================================
# Endpoints de Generación
# =============================================================================

@router.post(
    "/generar/{proyecto_id}",
    response_model=EstructuraEIAResponse,
    summary="Generar estructura EIA para un proyecto",
    description="""
    Genera la estructura completa del EIA para un proyecto específico.

    La estructura incluye:
    - **11 capítulos** según Art. 18 DS 40, adaptados al tipo de proyecto
    - **Lista de PAS** (Permisos Ambientales Sectoriales) aplicables
    - **Anexos técnicos** requeridos según características
    - **Plan de línea base** con componentes ambientales a caracterizar
    - **Estimación de complejidad** (nivel, tiempo, recursos, riesgos)

    Si ya existe una estructura, se puede forzar regeneración con `forzar=true`.
    """,
)
async def generar_estructura(
    proyecto_id: int,
    forzar: bool = Query(False, description="Forzar regeneración si ya existe"),
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> EstructuraEIAResponse:
    """Genera la estructura del EIA para un proyecto."""
    logger.info(f"Generando estructura EIA para proyecto {proyecto_id}")

    resultado = await servicio.generar_estructura(
        proyecto_id=proyecto_id,
        forzar_regenerar=forzar,
    )

    if not resultado:
        raise HTTPException(
            status_code=400,
            detail="No se pudo generar la estructura. Verifique que el proyecto existe y tiene tipo asignado."
        )

    return resultado


# =============================================================================
# Endpoints de Consulta de Estructura
# =============================================================================

@router.get(
    "/proyecto/{proyecto_id}",
    response_model=EstructuraEIAResponse,
    summary="Obtener estructura EIA de un proyecto",
    description="Retorna la estructura EIA más reciente del proyecto.",
)
async def get_estructura(
    proyecto_id: int,
    version: Optional[int] = Query(None, description="Versión específica (por defecto la última)"),
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> EstructuraEIAResponse:
    """Obtiene la estructura EIA de un proyecto."""
    estructura = await servicio.get_estructura_proyecto(proyecto_id, version)

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No existe estructura EIA para el proyecto {proyecto_id}"
        )

    return await servicio._convertir_a_response(estructura)


@router.get(
    "/proyecto/{proyecto_id}/resumen",
    response_model=ResumenEstructuraParaAsistente,
    summary="Obtener resumen de estructura para asistente",
    description="Retorna un resumen de la estructura formateado para el asistente IA.",
)
async def get_resumen_estructura(
    proyecto_id: int,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> ResumenEstructuraParaAsistente:
    """Obtiene resumen de estructura para el asistente."""
    return await servicio.get_resumen_para_asistente(proyecto_id)


# =============================================================================
# Endpoints de Capítulos
# =============================================================================

@router.get(
    "/proyecto/{proyecto_id}/capitulos",
    response_model=List[CapituloConEstado],
    summary="Obtener capítulos del EIA",
    description="Retorna la lista de capítulos del EIA con su estado y progreso.",
)
async def get_capitulos(
    proyecto_id: int,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> List[CapituloConEstado]:
    """Obtiene los capítulos del EIA de un proyecto."""
    estructura = await servicio.get_estructura_proyecto(proyecto_id)

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No existe estructura EIA para el proyecto {proyecto_id}"
        )

    return [CapituloConEstado(**c) for c in (estructura.capitulos or [])]


@router.patch(
    "/proyecto/{proyecto_id}/capitulo/{numero}",
    response_model=CapituloConEstado,
    summary="Actualizar estado de un capítulo",
    description="Actualiza el estado y progreso de un capítulo específico.",
)
async def actualizar_capitulo(
    proyecto_id: int,
    numero: int,
    datos: ActualizarCapituloRequest,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> CapituloConEstado:
    """Actualiza el estado de un capítulo."""
    estructura = await servicio.actualizar_estado_capitulo(
        proyecto_id=proyecto_id,
        capitulo_numero=numero,
        estado=datos.estado.value,
        progreso=datos.progreso_porcentaje,
    )

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No se encontró el capítulo {numero} del proyecto {proyecto_id}"
        )

    # Buscar el capítulo actualizado
    for cap in estructura.capitulos:
        if cap.get("numero") == numero:
            return CapituloConEstado(**cap)

    raise HTTPException(status_code=404, detail=f"Capítulo {numero} no encontrado")


# =============================================================================
# Endpoints de PAS (Permisos Ambientales Sectoriales)
# =============================================================================

@router.get(
    "/proyecto/{proyecto_id}/pas",
    response_model=List[PASConEstado],
    summary="Obtener PAS requeridos",
    description="Retorna la lista de PAS identificados con su estado de tramitación.",
)
async def get_pas(
    proyecto_id: int,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> List[PASConEstado]:
    """Obtiene los PAS requeridos de un proyecto."""
    estructura = await servicio.get_estructura_proyecto(proyecto_id)

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No existe estructura EIA para el proyecto {proyecto_id}"
        )

    return [PASConEstado(**p) for p in (estructura.pas_requeridos or [])]


@router.patch(
    "/proyecto/{proyecto_id}/pas/{articulo}",
    response_model=PASConEstado,
    summary="Actualizar estado de un PAS",
    description="Actualiza el estado de tramitación de un PAS específico.",
)
async def actualizar_pas(
    proyecto_id: int,
    articulo: int,
    datos: ActualizarPASRequest,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> PASConEstado:
    """Actualiza el estado de un PAS."""
    estructura = await servicio.actualizar_estado_pas(
        proyecto_id=proyecto_id,
        articulo=articulo,
        estado=datos.estado.value,
    )

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No se encontró el PAS Art. {articulo} del proyecto {proyecto_id}"
        )

    # Buscar el PAS actualizado
    for pas in estructura.pas_requeridos:
        if pas.get("articulo") == articulo:
            return PASConEstado(**pas)

    raise HTTPException(status_code=404, detail=f"PAS Art. {articulo} no encontrado")


# =============================================================================
# Endpoints de Anexos
# =============================================================================

@router.get(
    "/proyecto/{proyecto_id}/anexos",
    response_model=List[AnexoRequerido],
    summary="Obtener anexos requeridos",
    description="Retorna la lista de anexos técnicos requeridos con su estado.",
)
async def get_anexos(
    proyecto_id: int,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> List[AnexoRequerido]:
    """Obtiene los anexos requeridos de un proyecto."""
    estructura = await servicio.get_estructura_proyecto(proyecto_id)

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No existe estructura EIA para el proyecto {proyecto_id}"
        )

    return [AnexoRequerido(**a) for a in (estructura.anexos_requeridos or [])]


@router.patch(
    "/proyecto/{proyecto_id}/anexo/{codigo}",
    response_model=AnexoRequerido,
    summary="Actualizar estado de un anexo",
    description="Actualiza el estado de elaboración de un anexo técnico.",
)
async def actualizar_anexo(
    proyecto_id: int,
    codigo: str,
    datos: ActualizarAnexoRequest,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> AnexoRequerido:
    """Actualiza el estado de un anexo."""
    estructura = await servicio.actualizar_estado_anexo(
        proyecto_id=proyecto_id,
        codigo=codigo,
        estado=datos.estado.value,
    )

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No se encontró el anexo {codigo} del proyecto {proyecto_id}"
        )

    # Buscar el anexo actualizado
    for anexo in estructura.anexos_requeridos:
        if anexo.get("codigo") == codigo:
            return AnexoRequerido(**anexo)

    raise HTTPException(status_code=404, detail=f"Anexo {codigo} no encontrado")


# =============================================================================
# Endpoints de Línea Base
# =============================================================================

@router.get(
    "/proyecto/{proyecto_id}/linea-base",
    response_model=List[ComponenteLineaBaseEnPlan],
    summary="Obtener plan de línea base",
    description="Retorna el plan de línea base con componentes ambientales a caracterizar.",
)
async def get_linea_base(
    proyecto_id: int,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> List[ComponenteLineaBaseEnPlan]:
    """Obtiene el plan de línea base de un proyecto."""
    estructura = await servicio.get_estructura_proyecto(proyecto_id)

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No existe estructura EIA para el proyecto {proyecto_id}"
        )

    return [ComponenteLineaBaseEnPlan(**c) for c in (estructura.plan_linea_base or [])]


# =============================================================================
# Endpoints de Complejidad
# =============================================================================

@router.get(
    "/proyecto/{proyecto_id}/complejidad",
    response_model=EstimacionComplejidad,
    summary="Obtener estimación de complejidad",
    description="""
    Retorna la estimación de complejidad del EIA.

    La estimación considera:
    - Triggers Art. 11 activados
    - Número de PAS requeridos
    - Componentes de línea base
    - Sensibilidad territorial
    - Escala del proyecto
    """,
)
async def get_complejidad(
    proyecto_id: int,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> EstimacionComplejidad:
    """Obtiene la estimación de complejidad de un proyecto."""
    estructura = await servicio.get_estructura_proyecto(proyecto_id)

    if not estructura:
        raise HTTPException(
            status_code=404,
            detail=f"No existe estructura EIA para el proyecto {proyecto_id}"
        )

    if not estructura.estimacion_complejidad:
        raise HTTPException(
            status_code=404,
            detail="La estructura no tiene estimación de complejidad"
        )

    return EstimacionComplejidad(**estructura.estimacion_complejidad)


# =============================================================================
# Endpoints de Configuración por Tipo
# =============================================================================

@router.get(
    "/tipos/{tipo_codigo}/capitulos",
    response_model=List[CapituloEIAResponse],
    summary="Obtener capítulos configurados para un tipo",
    description="Retorna los capítulos EIA configurados para un tipo de proyecto.",
)
async def get_capitulos_tipo(
    tipo_codigo: str,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> List[CapituloEIAResponse]:
    """Obtiene los capítulos configurados para un tipo de proyecto."""
    capitulos = await servicio.get_capitulos_tipo(tipo_codigo)

    if not capitulos:
        raise HTTPException(
            status_code=404,
            detail=f"No hay capítulos configurados para el tipo '{tipo_codigo}'"
        )

    return [
        CapituloEIAResponse(
            id=c.id,
            tipo_proyecto_id=c.tipo_proyecto_id,
            numero=c.numero,
            titulo=c.titulo,
            descripcion=c.descripcion,
            contenido_requerido=c.contenido_requerido or [],
            es_obligatorio=c.es_obligatorio,
            orden=c.orden,
            activo=c.activo,
        )
        for c in capitulos
    ]


@router.get(
    "/tipos/{tipo_codigo}/linea-base",
    response_model=List[ComponenteLineaBaseResponse],
    summary="Obtener componentes de línea base para un tipo",
    description="Retorna los componentes de línea base configurados para un tipo de proyecto.",
)
async def get_componentes_tipo(
    tipo_codigo: str,
    servicio: EstructuraEIAService = Depends(get_servicio),
) -> List[ComponenteLineaBaseResponse]:
    """Obtiene los componentes de línea base configurados para un tipo."""
    componentes = await servicio.get_componentes_linea_base(tipo_codigo)

    if not componentes:
        raise HTTPException(
            status_code=404,
            detail=f"No hay componentes de línea base configurados para el tipo '{tipo_codigo}'"
        )

    return [
        ComponenteLineaBaseResponse(
            id=c.id,
            tipo_proyecto_id=c.tipo_proyecto_id,
            subtipo_id=c.subtipo_id,
            capitulo_numero=c.capitulo_numero,
            codigo=c.codigo,
            nombre=c.nombre,
            descripcion=c.descripcion,
            metodologia=c.metodologia,
            variables_monitorear=c.variables_monitorear or [],
            estudios_requeridos=c.estudios_requeridos or [],
            duracion_estimada_dias=c.duracion_estimada_dias,
            es_obligatorio=c.es_obligatorio,
            orden=c.orden,
            activo=c.activo,
        )
        for c in componentes
    ]

File: frontend/src/stores/estructuraEia.ts
/**
 * Store Pinia para Estructura EIA (Fase 2).
 */
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { get, post, patch } from '@/services/api'
import type {
  EstructuraEIA,
  CapituloConEstado,
  PASConEstado,
  AnexoRequerido,
  ComponenteLineaBaseEnPlan,
  EstimacionComplejidad,
  ActualizarCapituloRequest,
  ActualizarPASRequest,
  ActualizarAnexoRequest,
} from '@/types'

export const useEstructuraEiaStore = defineStore('estructuraEia', () => {
  // ==========================================================================
  // Estado
  // ==========================================================================

  const proyectoId = ref<number | null>(null)
  const estructura = ref<EstructuraEIA | null>(null)
  const cargando = ref(false)
  const generando = ref(false)
  const guardando = ref(false)
  const error = ref<string | null>(null)

  // ==========================================================================
  // Computed
  // ==========================================================================

  const tieneEstructura = computed(() => estructura.value !== null)

  const capitulos = computed<CapituloConEstado[]>(() =>
    estructura.value?.capitulos ?? []
  )

  const pasRequeridos = computed<PASConEstado[]>(() =>
    estructura.value?.pas_requeridos ?? []
  )

  const anexos = computed<AnexoRequerido[]>(() =>
    estructura.value?.anexos_requeridos ?? []
  )

  const planLineaBase = computed<ComponenteLineaBaseEnPlan[]>(() =>
    estructura.value?.plan_linea_base ?? []
  )

  const estimacion = computed<EstimacionComplejidad | null>(() =>
    estructura.value?.estimacion_complejidad ?? null
  )

  const progresoGeneral = computed(() => {
    if (!capitulos.value.length) return 0
    const completados = capitulos.value.filter(c => c.estado === 'completado').length
    return Math.round((completados / capitulos.value.length) * 100)
  })

  const capitulosCompletados = computed(() =>
    capitulos.value.filter(c => c.estado === 'completado').length
  )

  const capitulosEnProgreso = computed(() =>
    capitulos.value.filter(c => c.estado === 'en_progreso').length
  )

  const capitulosPendientes = computed(() =>
    capitulos.value.filter(c => c.estado === 'pendiente').length
  )

  const pasAprobados = computed(() =>
    pasRequeridos.value.filter(p => p.estado === 'aprobado').length
  )

  const pasEnTramite = computed(() =>
    pasRequeridos.value.filter(p => p.estado === 'en_tramite').length
  )

  const pasPendientes = computed(() =>
    pasRequeridos.value.filter(p =>
      p.estado === 'identificado' || p.estado === 'requerido'
    ).length
  )

  const anexosCompletados = computed(() =>
    anexos.value.filter(a => a.estado === 'completado').length
  )

  const anexosEnElaboracion = computed(() =>
    anexos.value.filter(a => a.estado === 'en_elaboracion').length
  )

  const duracionTotalLineaBase = computed(() =>
    planLineaBase.value.reduce((acc, c) => acc + (c.duracion_estimada_dias || 0), 0)
  )

  const nivelComplejidad = computed(() =>
    estimacion.value?.nivel ?? null
  )

  const tiempoEstimadoMeses = computed(() =>
    estimacion.value?.tiempo_estimado_meses ?? null
  )

  // ==========================================================================
  // Acciones - Cargar
  // ==========================================================================

  /**
   * Carga la estructura EIA del proyecto.
   */
  async function cargarEstructura(id: number): Promise<EstructuraEIA | null> {
    cargando.value = true
    error.value = null
    proyectoId.value = id

    try {
      estructura.value = await get<EstructuraEIA>(
        `/estructura-eia/proyecto/${id}`
      )
      return estructura.value
    } catch (e) {
      // 404 significa que no hay estructura aun
      if ((e as { status_code?: number }).status_code === 404) {
        estructura.value = null
        return null
      }
      error.value = e instanceof Error ? e.message : 'Error cargando estructura'
      throw e
    } finally {
      cargando.value = false
    }
  }

  /**
   * Recarga la estructura actual.
   */
  async function recargarEstructura(): Promise<void> {
    if (proyectoId.value) {
      await cargarEstructura(proyectoId.value)
    }
  }

  // ==========================================================================
  // Acciones - Generar
  // ==========================================================================

  /**
   * Genera la estructura EIA para un proyecto.
   */
  async function generarEstructura(
    id: number,
    forzarRegenerar: boolean = false
  ): Promise<EstructuraEIA | null> {
    generando.value = true
    error.value = null
    proyectoId.value = id

    try {
      estructura.value = await post<EstructuraEIA>(
        `/estructura-eia/generar/${id}`,
        { forzar_regenerar: forzarRegenerar }
      )
      return estructura.value
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Error generando estructura'
      throw e
    } finally {
      generando.value = false
    }
  }

  // ==========================================================================
  // Acciones - Actualizar Capitulos
  // ==========================================================================

  /**
   * Actualiza el estado de un capitulo.
   */
  async function actualizarCapitulo(
    numero: number,
    data: ActualizarCapituloRequest
  ): Promise<void> {
    if (!proyectoId.value) return

    guardando.value = true
    try {
      await patch(
        `/estructura-eia/proyecto/${proyectoId.value}/capitulo/${numero}`,
        data
      )

      // Actualizar localmente
      const capitulo = capitulos.value.find(c => c.numero === numero)
      if (capitulo) {
        capitulo.estado = data.estado
        if (data.progreso_porcentaje !== undefined) {
          capitulo.progreso_porcentaje = data.progreso_porcentaje
        }
        if (data.notas !== undefined) {
          capitulo.notas = data.notas
        }
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Error actualizando capitulo'
      throw e
    } finally {
      guardando.value = false
    }
  }

  // ==========================================================================
  // Acciones - Actualizar PAS
  // ==========================================================================

  /**
   * Actualiza el estado de un PAS.
   */
  async function actualizarPAS(
    articulo: number,
    data: ActualizarPASRequest
  ): Promise<void> {
    if (!proyectoId.value) return

    guardando.value = true
    try {
      await patch(
        `/estructura-eia/proyecto/${proyectoId.value}/pas/${articulo}`,
        data
      )

      // Actualizar localmente
      const pas = pasRequeridos.value.find(p => p.articulo === articulo)
      if (pas) {
        pas.estado = data.estado
        if (data.fecha_limite !== undefined) {
          pas.fecha_limite = data.fecha_limite
        }
        if (data.notas !== undefined) {
          pas.notas = data.notas
        }
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Error actualizando PAS'
      throw e
    } finally {
      guardando.value = false
    }
  }

  // ==========================================================================
  // Acciones - Actualizar Anexos
  // ==========================================================================

  /**
   * Actualiza el estado de un anexo.
   */
  async function actualizarAnexo(
    codigo: string,
    data: ActualizarAnexoRequest
  ): Promise<void> {
    if (!proyectoId.value) return

    guardando.value = true
    try {
      await patch(
        `/estructura-eia/proyecto/${proyectoId.value}/anexo/${codigo}`,
        data
      )

      // Actualizar localmente
      const anexo = anexos.value.find(a => a.codigo === codigo)
      if (anexo) {
        anexo.estado = data.estado
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Error actualizando anexo'
      throw e
    } finally {
      guardando.value = false
    }
  }

  // ==========================================================================
  // Helpers
  // ==========================================================================

  /**
   * Obtiene un capitulo por numero.
   */
  function getCapitulo(numero: number): CapituloConEstado | undefined {
    return capitulos.value.find(c => c.numero === numero)
  }

  /**
   * Obtiene un PAS por articulo.
   */
  function getPAS(articulo: number): PASConEstado | undefined {
    return pasRequeridos.value.find(p => p.articulo === articulo)
  }

  /**
   * Obtiene un anexo por codigo.
   */
  function getAnexo(codigo: string): AnexoRequerido | undefined {
    return anexos.value.find(a => a.codigo === codigo)
  }

  /**
   * Obtiene componentes de linea base obligatorios.
   */
  function getComponentesObligatorios(): ComponenteLineaBaseEnPlan[] {
    return planLineaBase.value.filter(c => c.es_obligatorio)
  }

  /**
   * Obtiene PAS obligatorios.
   */
  function getPASObligatorios(): PASConEstado[] {
    return pasRequeridos.value.filter(p => p.obligatoriedad === 'obligatorio')
  }

  // ==========================================================================
  // Limpiar
  // ==========================================================================

  function limpiar(): void {
    proyectoId.value = null
    estructura.value = null
    error.value = null
  }

  // ==========================================================================
  // Return
  // ==========================================================================

  return {
    // Estado
    proyectoId,
    estructura,
    cargando,
    generando,
    guardando,
    error,

    // Computed
    tieneEstructura,
    capitulos,
    pasRequeridos,
    anexos,
    planLineaBase,
    estimacion,
    progresoGeneral,
    capitulosCompletados,
    capitulosEnProgreso,
    capitulosPendientes,
    pasAprobados,
    pasEnTramite,
    pasPendientes,
    anexosCompletados,
    anexosEnElaboracion,
    duracionTotalLineaBase,
    nivelComplejidad,
    tiempoEstimadoMeses,

    // Acciones
    cargarEstructura,
    recargarEstructura,
    generarEstructura,
    actualizarCapitulo,
    actualizarPAS,
    actualizarAnexo,

    // Helpers
    getCapitulo,
    getPAS,
    getAnexo,
    getComponentesObligatorios,
    getPASObligatorios,

    // Limpiar
    limpiar,
  }
})
